src/main/java/com/onegini/examples/resourcegateway/web/ResourcesController.java

```diff
@@ -1,80 +1,98 @@
+  @GetMapping(value = "/application-details")
+  public ResponseEntity<ApplicationDetails> getApplicationDetails(@RequestHeader(AUTHORIZATION) final String authorizationHeader) {
+    final TokenIntrospectionResult tokenIntrospectionResult = getTokenIntrospectionResultFromHeader(authorizationHeader);
+    validateScopeAndTokenType(tokenIntrospectionResult, SCOPE_APPLICATION_DETAILS);
 
     final ApplicationDetails applicationDetails = new ApplicationDetails(tokenIntrospectionResult.getAppIdentifier(), tokenIntrospectionResult.getAppPlatform(),
         tokenIntrospectionResult.getAppVersion());
 
-    return new ResponseEntity<Object>(applicationDetails, OK);
+    return new ResponseEntity<>(applicationDetails, OK);
   }
 
-  @RequestMapping(value = "/user-id-decorated", method = RequestMethod.GET)
-  public ResponseEntity<?> getDecoratedUserId(@RequestHeader(AUTHORIZATION) final String authorizationHeader) {
-
-    final String accessToken = accessTokenExtractor.extractFromHeader(authorizationHeader);
-    final TokenIntrospectionResult tokenIntrospectionResult = tokenIntrospectionService.introspectAccessToken(accessToken);
+  @GetMapping(value = "/user-id-decorated")
+  public ResponseEntity<DecoratedUser> getDecoratedUserId(@RequestHeader(AUTHORIZATION) final String authorizationHeader) {
+    final TokenIntrospectionResult tokenIntrospectionResult = getTokenIntrospectionResultFromHeader(authorizationHeader);
 
     tokenTypeValidationService.validateImplicitAuthenticationToken(tokenIntrospectionResult.getAmr());
 
-    final DecoratedUserId decoratedUserId = new DecoratedUserIdBuilder()
+    final DecoratedUser decoratedUser = new DecoratedUserIdBuilder()
         .withUserId(tokenIntrospectionResult.getSub())
         .build();
 
-    return new ResponseEntity<>(decoratedUserId, OK);
+    return new ResponseEntity<>(decoratedUser, OK);
+  }
+
+  @PostMapping(value = "upload")
+  public ResponseEntity<MultipartResponse> upload(@RequestHeader(AUTHORIZATION) final String authorizationHeader,
+                                                  @ModelAttribute final FormDataWithFiles formDataWithFiles) {
+    final TokenIntrospectionResult tokenIntrospectionResult = getTokenIntrospectionResultFromHeader(authorizationHeader);
+    validateScopeAndTokenType(tokenIntrospectionResult, SCOPE_WRITE);
+
+    final MultipartResponse multipartResponse = multipartService.parse(formDataWithFiles);
+
+    return new ResponseEntity<>(multipartResponse, OK);
+  }
+
+  private void validateScopeAndTokenType(final TokenIntrospectionResult tokenIntrospectionResult, final String requiredScope) {
+    scopeValidationService.validateScopeGranted(tokenIntrospectionResult.getScope(), requiredScope);
+    tokenTypeValidationService.validateNoImplicitAuthenticationToken(tokenIntrospectionResult.getAmr());
+  }
+
+  private TokenIntrospectionResult getTokenIntrospectionResultFromHeader(final String authorizationHeader) {
```

Comment:
I've found out that in case the authorization header is missing the service responds with 400 while I believe it should return 401. Not necessarily related to your change so I guess we can leave it as is.

Reply:
It was an easy fix so I did it

Reply:
Cool 💪  Thanks!